### ISSUE TITLE
about some technical details

### ISSUE THREAD

**Issue by Zhaocanaxx** [dimitrismallis/KeypointsToLandmarks#4]
Hi, sorry to bother you. I think your work is really meaningful but I have some questions. 
question 1: how to get the ground-truth heatmap for landmarks?
question 2: how to connect  each landmark and each descriptor? because the two come from different head network.

**Comment by dimitrismallis** [comment]
Hi and thank you for you interest in this work. Regarding your questions.

1) This is an unsupervised landmark discovery method so throughout the method we use pseudo-annotations. In particular, the pseudo-annotations used to construct training heatmaps are extracted from the model itself at:
 https://github.com/dimitrismallis/KeypointsToLandmarks/blob/3b182316bb1b946a0894a9c94072e6f79b0a5712/FanClass.py#L315 

   For bootstrapping the method, initial heatmaps are constructed from landmark pseudo-labels derived from the keypoint detector SuperPoint in:
 https://github.com/dimitrismallis/KeypointsToLandmarks/blob/3b182316bb1b946a0894a9c94072e6f79b0a5712/SuperPoint.py#L49

2) Indeed the landmark keypoints and descriptors are produced by separate output heads. The description head yields a dense output volume containing a descriptor for every pixel. We first extract the landmark keypoints from the detection head, then gather the corresponding descriptors at those keypoint locations. See the the function below for for sampling of output landmark descriptors. It is using pytorch's [grid_sample](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html).https://github.com/dimitrismallis/KeypointsToLandmarks/blob/3b182316bb1b946a0894a9c94072e6f79b0a5712/utils.py#L189
